### DISCUSSION TITLE
Self made thermostat

### DISCUSSION THREAD

**Discussion by fuecy** [Arnold-n/P1P2MQTT#137] [Show and tell]
Hi,

I just want to receive some feedback on my thermostat and check if it makes sence to you.

I have a daikin E series but without the standard out-of-the box thermostat. Instead i have build my own thermostat on a arduino uno.
Iam using the relays of my daikin to control hot and cold demand and iam using P1P2MQTT to control my hot water. Since i have a 300L tank, i heat it only once on the most cheapest hour of the day (and there was not enough solar power to do otherwise)

So basically i have made a (a very simple linear) model that predicts my heat demand for the day based on todays and tomorrows min and max temperature and solar prediction and it has a correction based on my average indoor temperature. Iam getting the values just from online weather API's and i have some temperature sensors connected for the room temperatures. (now still on a service in python but planning to integrated it into the arduino)

Once i have the time prediction, i matches that with the cheapest hours of my dynamic energy contract plus i give some preference for the night hours so that my heat pump is on for long runs in cheap hours.

I didn’t look into COP and other efficiency gains, iam lagging the knowledge of those. Maybe you can give me some recommendations.

Does this make sense to you, would you recommend taking a complete different approach like running all day on lower power settings? As i assume now when the heat demand is enabled, it goes to max. of the internal daikin weather depended water temperature.

<img width="353" alt="demand" src="https://github.com/user-attachments/assets/3891968c-dc42-4b04-81d8-56421103a9be">

Relay schedule, 5 hours on the cheapest price (note its a exact graph from 0 to 24 and the straight line is just because its future)

<img width="396" alt="model" src="https://github.com/user-attachments/assets/56937656-0cd0-43de-bca5-f2fca0c7a069">

Model outcome in seconds. (the outcome is the sum of the below intimidate values)

<img width="352" alt="buizen" src="https://github.com/user-attachments/assets/4434605d-665f-41b9-8ecd-e20123169856">

Pipe temperatures (with 1 defrost) and hot water enabled due to enough solar energy

<img width="357" alt="energy prices" src="https://github.com/user-attachments/assets/76cfa82d-4ab1-4daa-b53f-2d1fe3d7ad05">

Energy prices and cost. (there is not that much difference in the hour prices today, but on average it makes a lot of difference)

**Comment by Arnold-n** [comment]
It depends on what you are aiming to minimize: heat pump wear, cost of energy, or CO2 emitted. And it may be difficult to relate power consumer to CO2 emitted as this will differ over the day. But in general your approach is sound for using energy when cheap (or priced negative) it seems environmentally friendly to me too. If you want to further mprove, I guess the best way is to (slowly, infrequently) modulate the LWT setpoint.

**Comment by fuecy** [comment]
Hi thanks, i will definitely look in to  the LWT setpoint. My experience now is, when i lower it, the compressor start turning on and off continuously. But i didnt investigate more yet.

Would it be possible to connect P1P2MQTT physical from another arduino to enable and disable hot water (not using network/mqtt). For example over serial or by a relays?

**Reply by Arnold-n** [reply]
> Hi thanks, i will definitely look in to the LWT setpoint. My experience now is, when i lower it, the compressor start turning on and off continuously. But i didnt investigate more yet.

Did you (if possible) increase overshoot [9-04] to 4 or try quiet mode 3 to reduce this?

> Would it be possible to connect P1P2MQTT physical from another arduino to enable and disable hot water (not using network/mqtt). For example over serial or by a relays?

This is strongly discouraged, as the P1P2MQTT bridge operates directly from the P1/P2 wires and no galvanic isolation is provided. In theory it would be possible via relays using the ethernet adapter socket which provides access to a few GPIO pins.

**Reply by fuecy** [reply]
overshoot was on 3, now its on 4 and i will try with silence mode, thanks!

Would be really nice to have some "configurable" smartgrids in a next hardware version (if you ever plan such a thing). There are a lot of solar panel vendors that provide an option to enable smartgrid relays based on energy price and/or solar power. The smartgrid options daikin provides are, at least for my headpump, limited. 

In my case i have solaredge, daikin provides a 3th party modbus solution (which solar edge is using) connected to the p1p2. But it feels not properly supported to me and more a edge case for them (beside its expensive and limited). With configurable smartgrids you extend the use cases for your awesome product. Sorry it feels like iam writing a sales pitch but i really thing it brings value.

**Reply by Arnold-n** [reply]
Thanks - V1.3 is just ready but next year a new version is planned. Would you have a reference to how your solaredge can connects (how many outputs, which specs)? If it offers a clean relays output (thus galvanically isolated) it would be relatively easy to connect to existing hardware (>=V1.1). Otherwise the bridge would have to ensure isolation.

**Reply by fuecy** [reply]
According to my [daikin documentation (9.3.11)](https://www.daikin.nl/content/dam/document-library/Installer-reference-guide/heat/air-to-water-heat-pump-low-temperature/erga-ev/ERGA-EV_EVH_EVH7_ERGA04-08EVA_EHBH_X-6V_9W_Installer%20reference%20guide_4PEN629085-1C_English.pdf), you have low voltage and high voltage contacts. I think most solar panel providers also support both however i see heat pump is mostly on low power contacts. Low voltage means its a relay and the device provides the power. High power means it 230v switched

As an example, with configurable i mean having more control over what the relay state normally does (see below). So basically having relay on the P1P2MQTT which trigger (an number of) write commands to the headpump.
<img width="602" alt="image" src="https://github.com/user-attachments/assets/55a26aff-87f3-4bb4-bfc0-8c33cbaba6e5">

Other notable references.
https://www.nibe.eu/nl-nl/producten/smart-home/smart-grid-ready
https://knowledge-center.solaredge.com/sites/kc/files/Smart_Energy_Management_Flyer_nld.pdf
https://www.sma-benelux.com/kenniscentrum/tech-tips-voor-de-installatie/sg-ready-warmtepompen.
https://alklima.nl/blog/slim-besparen-koppel-je-zonnepanelen-aan-je-warmtepomp-voor-een-smart-grid-ready-toekomst

please note that it seems that [eebus](https://www.eebus.org/solutions/#use_case_tab) is taking over smart grid or at least i see some vendors moving in that direction with their latest products

**Comment by dlefol** [comment]
Hi,

I read with interest your post and I think your logic is sound and I use a similar approach myself but I have a couple of questions / comments:

1. I don't really understand why you are using an Arduino to implement your logic. By the look of your graphs you have the P1P2MQTT bridge connected to Home Assistant. If so why not just use the climate entities provided by the bridge to directly put the heating / cooling on or off ? I have set up my heat pump to run based on outside temperature only and I start and stop the heating using various criteria (indoor temperature, sun, minimum run time, ...) by changing the climate entity of P1P2MQTT from off to heat via HA (using an appdaemon code I wrote to implement the logic I wanted).
2. Concerning the optimisation of COP, the way I do it (not sure it's the best way but seems to work ok) is to reduce the LWT by -1 when the heat pump start (I noticed the HP tends to start at full power and I want to avoid this) and then I increase the LWT by 1 each time the pump speed drops below 10 L/min and the COP is below 5 to a maximum of LWT + 2 (basically I try to force the HP to always work with a high flow rate at medium power as it seems to provide the best COP). I have UFH everywhere and my LWT target temperature is usually around 26 degrees at the moment since it's not so cold so this logic might not work for your particular case.

Thanks Arnold at the same time for this great project which enables so much freedom to control the HP :)

**Reply by savage7** [reply]
@dlefol That sounds like a very interesting approach! 
 
Is it safe to use this much write operations to the heat pump? Reading the documentation I was thinking that writing the EEPROM should be minimized.

**Reply by dlefol** [reply]
I would think that it's not a problem but I don't really know to be honest.
I just checked and my write budget is most of the time at 100 (max) and I have around 20-40 write operations spread through the day which seems reasonable to me.
Moreover most of these operations (starting/stopping heating and changing temperature) can be done via the onecta app and Daikin does not seem to warn not too use it too much in the app so I guess it's safe (at least I hope so!)

**Reply by Arnold-n** [reply]
There is only one warning by Daikin in one specific older manual for one F-series model airco, which was likely so simple it didn't implement wear leveling. I think the warning is not really needed any more, but just to be sure I kept the write budget in, with an option to switch it off. I have never seen any warning from Daikin related to the Onecta app, LAN adapter, or Daikin Home Hub either.

**Reply by fuecy** [reply]
Thank you for the feedback and insights!

I'm using an Arduino instead of relying entirely on my server hardware because I prefer to keep critical functions separate from my main server. Previously, I ran a similar automation setup through Home Assistant, similar to what you’re describing.

Even though my server hardware is robust, I’d rather avoid the risk of mid-winter hardware issues that would require a full recovery process—and in the worst case, leave me without heating. Investing in fully redundant server hardware isn’t justifiable for me financially, especially given that an Arduino is easy to replace, and I have spares on hand. But you are right when you say, I could also take a pi with a second home assistant, that actually would be worth looking in to. Thanks! 

Additionally, my home doesn’t have a traditional thermostat, so I'm aiming for a solution that’s self-contained and could be left behind if I ever move. While I’m currently using APIs for certain controls, my plan includes building a fallback setup using wired indoor and outdoor temperature sensors and controls.

Regarding the write operations, I’m under the impression that non-persistent changes (those not requiring a restart) remain in memory rather than being stored on the EEPROM. To test this, I ran a few trials by cutting the power and observing the state upon reboot, which confirmed this assumption.

For COP optimization, I find your approach intriguing and would like to explore it further. My heat pump also starts with a high flow rate as you can see, and I suspect there may be an intended reason behind this, as it’s part of the default behavior from Daikin. However, I’m currently tuning my model based on the LWT data I’ve accumulated, so I’ll need to wait before experimenting with new COP adjustments.

**Comment by FlexxFR** [comment]
Interesting topic and I'm trying to do similar things in HA. (like suggested earlier in this thread).
My approach was a bit simpler and just altering the setpoint a little base on energy price. (Maybe this is too simple?)

However, last night the cheapest hours where during night-time and for the first time this year the heatpump needed to make use of the BUH almost full time. (so not only for a defrosting cycle). Of course, it does not help to use double amount of power while it's only 10-20% cheaper :) :) 

![image](https://github.com/user-attachments/assets/aa0f21c3-b04b-4e1e-82c4-6addf49da228)

I assume this has something to do with the outdoor temp. However, I can't find anywhere what the threshold is, if this could be adjusted and if so, is it wise to do so?  Daikin measured an outdoor temp of 3.8C when the heating cycle started, but dropped to 2, even 1 during the entire cycle.

**Reply by fuecy** [reply]
Based on your data i can only share my experience. For me, my outdoor temperature is influenced by the unit itself but not that much. I have 2 other outdoor temperature sensors and i never noticed huge gaps. I am looking for the hc_sensors_temperature_refrigerant which in my case triggers the defrost around -2. I also noticed that sometimes this happens a few times till it find a stable -1.5 temperature.
One more thing i can think off, i have a backup heating system, not sure about yours that starts to work when the HP cant. No settings iam aware of for this beside disabling it fully.

i started to alter my setpoint logic also. Now i start increasing the setpoint when the prices are cheap and even more when they get close to 0 (in case of storm) or when i have too much solar power. i dont want to sell my solar, i want to use it myself. Then my temperature in the house raises but the next day, it corrects by reducing the HP operation time.

i dont have a real time power meter on my HP, thats going to be the next thing which allows to find the most efficient combinations.

**Reply by fuecy** [reply]
maybe to add, my backup heating (BUH1_Q) is on (only) when my hot water is heating.

**Reply by FlexxFR** [reply]
My situation improved a lot by tweaking the "silent mode", it's now back on "full" and maybe this enables more heating power from the HP itself, without BUH.

For example, a few nights it was even colder that my previous example (~ 5) but the HP worked fine, BUH was only needed for defrost cycles. (red spikes)
![image](https://github.com/user-attachments/assets/0578ddeb-f27f-4bc7-956a-0c194807deeb)

However, due to an unrelated failure my HP was disabled for 20 hours before I noticed, so the house was really cooled down, all floors. During day-time I restarted the HP, which started heating but, probably due to the low temp state, it could reach the SP LVT. So it enabled the BUH to do the best is can.

Made me wonder, why? I mean, no problem if it takes just a bit longer. Is it possible to enable a "do the best you can mode without using BUH"?. There are still hours, even cheap ones, that the HP is doing nothing.

completely disable BUH is not possible as it's required for defrost & DHW legionella run.
